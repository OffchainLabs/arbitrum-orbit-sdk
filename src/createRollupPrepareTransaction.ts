import { Transaction, decodeFunctionData } from 'viem';

import { rollupCreator } from './contracts';
import { CreateRollupFunctionInputs } from './types/createRollupTypes';

function createRollupDecodeFunctionData(data: `0x${string}`) {
  return decodeFunctionData({
    abi: rollupCreator.abi,
    data,
  });
}

export type CreateRollupTransaction = Transaction & {
  getInputs(): CreateRollupFunctionInputs;
};

export function createRollupPrepareTransaction(tx: Transaction): CreateRollupTransaction {
  return {
    ...tx,
    getInputs: function () {
      const { functionName, args } = createRollupDecodeFunctionData(tx.input);

      if (functionName !== 'createRollup') {
        throw new Error(`
          [createRollupPrepareTransaction] expected function name to be "createRollup" but got "${functionName}"
        `);
      }

      if (typeof args === 'undefined') {
        throw new Error(`[createRollupPrepareTransaction] failed to decode function data`);
      }

      return args;
    },
  };
}
