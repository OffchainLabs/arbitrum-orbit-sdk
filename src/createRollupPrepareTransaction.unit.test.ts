import { it, expect } from 'vitest';
import { Transaction } from 'viem';

import { createRollupPrepareTransaction } from './createRollupPrepareTransaction';

// https://sepolia.arbiscan.io/tx/0x5b0b49e0259289fc89949a55a5ad35a8939440a55065d29b14e5e7ef7494efff
it('successfully parses input calldata for a chain created with RollupCreator v1.1', () => {
  const { getInputs } = createRollupPrepareTransaction({
    input:
      '0xcb73d6e2000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000e00000000000000000000000003c3a5b44fab0e2025160a765348c21c08e41d1af00000000000000000000000000000000000000000000000000000000000005400000000000000000000000000000000000000000000000000000000000019999000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000005f5e100000000000000000000000000000000000000000000000000000000000000009600000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000016345785d8a00008b104a2e80ac6165dc58b9048de12f301d70b02a0ab51396c22b4b4b802a16a400000000000000000000000041c966f99de0ca6f6531fbcac9db7eabdf119744000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000006fee560eb00000000000000000000000000000000000000000000000000000000000001c000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007080000000000000000000000000000000000000000000000000000000000000012c00000000000000000000000000000000000000000000000000000000000546000000000000000000000000000000000000000000000000000000000000000e1000000000000000000000000000000000000000000000000000000000000002757b22686f6d657374656164426c6f636b223a302c2264616f466f726b426c6f636b223a6e756c6c2c2264616f466f726b537570706f7274223a747275652c22656970313530426c6f636b223a302c2265697031353048617368223a22307830303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030222c22656970313535426c6f636b223a302c22656970313538426c6f636b223a302c2262797a616e7469756d426c6f636b223a302c22636f6e7374616e74696e6f706c65426c6f636b223a302c2270657465727362757267426c6f636b223a302c22697374616e62756c426c6f636b223a302c226d756972476c6163696572426c6f636b223a302c226265726c696e426c6f636b223a302c226c6f6e646f6e426c6f636b223a302c22636c69717565223a7b22706572696f64223a302c2265706f6368223a307d2c22617262697472756d223a7b22456e61626c654172624f53223a747275652c22416c6c6f774465627567507265636f6d70696c6573223a66616c73652c2244617461417661696c6162696c697479436f6d6d6974746565223a747275652c22496e697469616c4172624f5356657273696f6e223a32302c2247656e65736973426c6f636b4e756d223a302c224d6178436f646553697a65223a32343537362c224d6178496e6974436f646553697a65223a34393135322c22496e697469616c436861696e4f776e6572223a22307834314339363666393944653063413646363533316662634163394462376561424446313139373434227d2c22636861696e4964223a33303034363234393139357d000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000000008e842599f71abd661737bb3108a53e5b1787c791',
  } as unknown as Transaction);

  expect(getInputs()).toMatchSnapshot();
});

// https://sepolia.arbiscan.io/tx/0x77db43157182a69ce0e6d2a0564d2dabb43b306d48ea7b4d877160d6a1c9b66d
it('successfully parses input calldata for a chain created with RollupCreator v2.1', () => {
  const { getInputs } = createRollupPrepareTransaction({
    input:
      '0x331f9b0b0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000005600000000000000000000000000000000000000000000000000000000000019999000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000005f5e10000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000009600000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000016345785d8a0000260f5fa5c3176a856893642e149cf128b5a8de9f828afec8d11184415dd8dc6900000000000000000000000031bac660fbb4ef1b91409db42cb4c6e75527d004000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007a91eb4ee00000000000000000000000000000000000000000000000000000000000001c000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000007080000000000000000000000000000000000000000000000000000000000000012c00000000000000000000000000000000000000000000000000000000000546000000000000000000000000000000000000000000000000000000000000000e1000000000000000000000000000000000000000000000000000000000000002757b22686f6d657374656164426c6f636b223a302c2264616f466f726b426c6f636b223a6e756c6c2c2264616f466f726b537570706f7274223a747275652c22656970313530426c6f636b223a302c2265697031353048617368223a22307830303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030303030222c22656970313535426c6f636b223a302c22656970313538426c6f636b223a302c2262797a616e7469756d426c6f636b223a302c22636f6e7374616e74696e6f706c65426c6f636b223a302c2270657465727362757267426c6f636b223a302c22697374616e62756c426c6f636b223a302c226d756972476c6163696572426c6f636b223a302c226265726c696e426c6f636b223a302c226c6f6e646f6e426c6f636b223a302c22636c69717565223a7b22706572696f64223a302c2265706f6368223a307d2c22617262697472756d223a7b22456e61626c654172624f53223a747275652c22416c6c6f774465627567507265636f6d70696c6573223a66616c73652c2244617461417661696c6162696c697479436f6d6d6974746565223a747275652c22496e697469616c4172624f5356657273696f6e223a32302c2247656e65736973426c6f636b4e756d223a302c224d6178436f646553697a65223a32343537362c224d6178496e6974436f646553697a65223a34393135322c22496e697469616c436861696e4f776e6572223a22307833316241433636306642423465463142393134303964423432434234433665373535323764303034227d2c22636861696e4964223a33323930323133323937347d00000000000000000000000000000000000000000000000000000000000000000000000000000000000004000000000000000000000000dbb2c9923b5de18b151bc55ed571edcd1fc7eeb900000000000000000000000084b2effdd54aa5dce861440ae9b9d320b043a64c00000000000000000000000039b4ce32e557225a401917a701ac4d267648635a000000000000000000000000e2d0cc872647b1a129b942bbfc980b31e8e94df200000000000000000000000000000000000000000000000000000000000000080000000000000000000000009464ad09ce0d157efc6c7eefe577c5aa6dec804d000000000000000000000000c02386b2d10a37e71145b6a7f5569e1db46042130000000000000000000000006c4483f6dcedb5a758b28d28e11f805bdacfe245000000000000000000000000f89836dc33f58394fb34b3e690d3829a65c74286000000000000000000000000d1d25aeb44e99b35e2d5e7db961c73ec94f92a24000000000000000000000000dc4eca04032b9178020ff0c6154a7e87309a429f000000000000000000000000e122f9838a4c8e6834f24d1b9dca92eb52a8e17e000000000000000000000000b280fd59090f3d95588d94bacd22b336de2278e0',
  } as unknown as Transaction);

  expect(getInputs()).toMatchSnapshot();
});
