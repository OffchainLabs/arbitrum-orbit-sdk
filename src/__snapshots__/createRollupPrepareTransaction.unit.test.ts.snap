// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`successfully parses input calldata for a chain created with RollupCreator v1.1 1`] = `
[
  {
    "batchPoster": "0x3C3A5b44FAB0e2025160a765348c21C08e41d1Af",
    "config": {
      "baseStake": 100000000000000000n,
      "chainConfig": "{\\"homesteadBlock\\":0,\\"daoForkBlock\\":null,\\"daoForkSupport\\":true,\\"eip150Block\\":0,\\"eip150Hash\\":\\"0x0000000000000000000000000000000000000000000000000000000000000000\\",\\"eip155Block\\":0,\\"eip158Block\\":0,\\"byzantiumBlock\\":0,\\"constantinopleBlock\\":0,\\"petersburgBlock\\":0,\\"istanbulBlock\\":0,\\"muirGlacierBlock\\":0,\\"berlinBlock\\":0,\\"londonBlock\\":0,\\"clique\\":{\\"period\\":0,\\"epoch\\":0},\\"arbitrum\\":{\\"EnableArbOS\\":true,\\"AllowDebugPrecompiles\\":false,\\"DataAvailabilityCommittee\\":true,\\"InitialArbOSVersion\\":20,\\"GenesisBlockNum\\":0,\\"MaxCodeSize\\":24576,\\"MaxInitCodeSize\\":49152,\\"InitialChainOwner\\":\\"0x41C966f99De0cA6F6531fbcAc9Db7eaBDF119744\\"},\\"chainId\\":30046249195}",
      "chainId": 30046249195n,
      "confirmPeriodBlocks": 150n,
      "extraChallengeTimeBlocks": 0n,
      "genesisBlockNum": 0n,
      "loserStakeEscrow": "0x0000000000000000000000000000000000000000",
      "owner": "0x41C966f99De0cA6F6531fbcAc9Db7eaBDF119744",
      "sequencerInboxMaxTimeVariation": {
        "delayBlocks": 28800n,
        "delaySeconds": 345600n,
        "futureBlocks": 300n,
        "futureSeconds": 3600n,
      },
      "stakeToken": "0x0000000000000000000000000000000000000000",
      "wasmModuleRoot": "0x8b104a2e80ac6165dc58b9048de12f301d70b02a0ab51396c22b4b4b802a16a4",
    },
    "deployFactoriesToL2": true,
    "maxDataSize": 104857n,
    "maxFeePerGasForRetryables": 100000000n,
    "nativeToken": "0x0000000000000000000000000000000000000000",
    "validators": [
      "0x8E842599F71ABD661737bb3108a53E5b1787c791",
    ],
  },
]
`;

exports[`successfully parses input calldata for a chain created with RollupCreator v2.1 1`] = `
[
  {
    "batchPosterManager": "0x0000000000000000000000000000000000000000",
    "batchPosters": [
      "0x9464aD09CE0d157Efc6c7EefE577C5AA6dEc804D",
      "0xC02386B2D10a37e71145B6A7f5569e1db4604213",
      "0x6C4483f6DCEDb5a758b28D28E11F805bdACFE245",
      "0xf89836dc33f58394fB34B3e690D3829A65C74286",
      "0xd1d25Aeb44E99B35E2d5E7Db961c73ec94f92a24",
      "0xDc4ECA04032b9178020Ff0c6154A7e87309a429f",
      "0xe122F9838A4C8e6834F24D1b9dCa92eb52a8E17e",
      "0xB280Fd59090f3D95588d94BACD22b336dE2278e0",
    ],
    "config": {
      "baseStake": 100000000000000000n,
      "chainConfig": "{\\"homesteadBlock\\":0,\\"daoForkBlock\\":null,\\"daoForkSupport\\":true,\\"eip150Block\\":0,\\"eip150Hash\\":\\"0x0000000000000000000000000000000000000000000000000000000000000000\\",\\"eip155Block\\":0,\\"eip158Block\\":0,\\"byzantiumBlock\\":0,\\"constantinopleBlock\\":0,\\"petersburgBlock\\":0,\\"istanbulBlock\\":0,\\"muirGlacierBlock\\":0,\\"berlinBlock\\":0,\\"londonBlock\\":0,\\"clique\\":{\\"period\\":0,\\"epoch\\":0},\\"arbitrum\\":{\\"EnableArbOS\\":true,\\"AllowDebugPrecompiles\\":false,\\"DataAvailabilityCommittee\\":true,\\"InitialArbOSVersion\\":20,\\"GenesisBlockNum\\":0,\\"MaxCodeSize\\":24576,\\"MaxInitCodeSize\\":49152,\\"InitialChainOwner\\":\\"0x31bAC660fBB4eF1B91409dB42CB4C6e75527d004\\"},\\"chainId\\":32902132974}",
      "chainId": 32902132974n,
      "confirmPeriodBlocks": 150n,
      "extraChallengeTimeBlocks": 0n,
      "genesisBlockNum": 0n,
      "loserStakeEscrow": "0x0000000000000000000000000000000000000000",
      "owner": "0x31bAC660fBB4eF1B91409dB42CB4C6e75527d004",
      "sequencerInboxMaxTimeVariation": {
        "delayBlocks": 28800n,
        "delaySeconds": 345600n,
        "futureBlocks": 300n,
        "futureSeconds": 3600n,
      },
      "stakeToken": "0x0000000000000000000000000000000000000000",
      "wasmModuleRoot": "0x260f5fa5c3176a856893642e149cf128b5a8de9f828afec8d11184415dd8dc69",
    },
    "deployFactoriesToL2": false,
    "maxDataSize": 104857n,
    "maxFeePerGasForRetryables": 100000000n,
    "nativeToken": "0x0000000000000000000000000000000000000000",
    "validators": [
      "0xDBb2c9923b5DE18B151bC55Ed571EDcd1fC7EeB9",
      "0x84B2EFFDd54aA5dce861440Ae9B9D320b043a64c",
      "0x39B4Ce32E557225a401917a701ac4d267648635a",
      "0xe2D0cC872647B1A129B942BbFC980B31E8e94Df2",
    ],
  },
]
`;
